### PR TITLE
Define four segments that search over all custom variable name/value columns instead of one per column

### LINKED ISSUE #6031: Make it easier to create Segments with Custom Variables
**Context**
Piwik lets you assign [Custom Variables](http://piwik.org/docs/custom-variables/) to visit and/or pageviews (and [site searches](http://piwik.org/docs/site-search/), [events](http://piwik.org/docs/event-tracking/), etc.). Users can then create [Custom Segments](http://piwik.org/docs/segmentation/) to get reports for all visits and pages matching a particular Custom Variable.

**Goal**
The goal of this ticket is to make creating segments with custom variables more user friendly.

**Why are segments hard to use with Custom variables?**

To create a segment with a custom variable, user has to know which slot the custom variable was created in (slot is a number from 1 to N). A segment to select only page views tagged with a particular author would like this: `customVariableNamePage1==author,customVariableValuePage1==xyz`

This is hard to use by users who may not know which slot was a given custom variable set.

**Proposed solution** 

 * Create new segments providing shortcuts, ie.
  * `customVariableNamePage==name_here` which would automatically select any page view with a custom variable name set with `name_here`. The generated SQL would look something like `custom_var_k1 == name_here OR custom_var_k2 == name_here OR custom_var_k3 == name_here OR custom_var_k4 == name_here OR custom_var_k5 == name_here`
  * similarly create new segment for the value `customVariableValuePage`
  * and the same segments for the "Visit" scope: `customVariableName` and `customVariableValue`

 * the auto-completer for the segment `customVariableName` would show all possible custom variable names for all slots. 

Adding this new "magic segment" matching any slot should make the act of creating custom segments with custom variables much easier. Thoughts?

### PR DESCRIPTION
**Pull request by diosmosis**
This PR creates four new segments: customVariableName, customVariableValue, customVariablePageName, customVariablePageValue. Each one searches over every custom variable slot so it is not necessary to manually string the different custom variable searches together (eg, `customVariableName1=...,customVariableName2=...,customVariableName3=...`).

To accomplish this, I modified the `Segment::setSqlSegment` field so it can be an array (though this isn't documented). If found, `Piwik\Segment\SegmentExpression` will automatically insert `OR`s between conditions matching the value on each specified table column.

To get the auto suggest API to work, I had to allow the suggestedValuesCallback property Segments to be non-strings and to allow them to accept a DataTable argument. If it does accept such an argument, the Live.getLastVisitDetails data (collected in API::getSuggestedValuesForSegment) is passed to the method.

Also includes a change of base classes for CustomVariableName/CustomVariableValue dimensions so they will be included in the `Dimension::getAllDimensions()` result. Not sure if this will cause side effects.

The build may not pass yet.

Fixes #6031

### CODE REVIEW

File: core/Segment/SegmentExpression.php [hunk 1]
```diff
@@ -253,21 +257,38 @@ protected function getSqlMatchFromDefinition($def, &$availableTables)
...
-            || is_null($value)
-        ) {
-            $sqlExpression = "( $field $sqlMatch )";
-        } else {
-            if ($alsoMatchNULLValues) {
-                $sqlExpression = "( $field IS NULL OR $field $sqlMatch ? )";
+        $sqlExpressions = array();
+        $values = array();
+        foreach ($fields as $field) {
+            $sqlMatchReplaced = str_replace('%s', $field, $sqlMatch);
```
Comment (tsteur) [inline core/Segment/SegmentExpression.php]: Can you check / make sure that no SQL injection is possible here? Not sure where `$fields` or `$def[0]` is coming from and whether they were validated against a whitelist or so.

**Comment by tsteur** [comment]
It is still possible to apply a segment for a specific one right? Can we maybe show a title in the UI that says that we're search in all custom variables for one segment and that results can be inaccurate if the search matches values from different custom variables? If I understand this feature right one has to be very careful using it, maybe even allowing only `==` comparison? Will comment this on the actual issue since it's not really related to the PR.

File: core/Segment/SegmentExpression.php [hunk 2]
```diff
@@ -253,21 +257,38 @@ protected function getSqlMatchFromDefinition($def, &$availableTables)
...
+
+            if ($matchType === self::MATCH_ACTIONS_CONTAINS
+                || is_null($value)
+            ) {
+                $sqlExpression = "( $sqlMatchReplaced )";
             } else {
-                $sqlExpression = "$field $sqlMatch ?";
+                if ($alsoMatchNULLValues) {
+                    $sqlExpression = "( $field IS NULL OR $sqlMatchReplaced ? )";
```
Comment (tsteur) [inline core/Segment/SegmentExpression.php]: Also here should check there is no possible SQL injection, as mentioned I don't know where `$field` is from etc

File: plugins/API/API.php [hunk 3]
```diff
@@ -520,12 +532,16 @@ public function getSuggestedValuesForSegment($segmentName, $idSite)
...
+            // Cleanup data to return the top suggested (non empty) labels for this segment
+            $values = $table->getColumn($segmentName);
 
-        // Select also flattened keys (custom variables "page" scope, page URLs for one visit, page titles for one visit)
-        $valuesBis = $table->getColumnsStartingWith($segmentName . ColumnDelete::APPEND_TO_COLUMN_NAME_TO_KEEP);
-        $values = array_merge($values, $valuesBis);
+            // Select also flattened keys (custom variables "page" scope, page URLs for one visit, page titles for one visit)
+            $valuesBis = $table->getColumnsStartingWith($segmentName . ColumnDelete::APPEND_TO_COLUMN_NAME_TO_KEEP);
+            $values = array_merge($values, $valuesBis);
+        }
```
Comment (tsteur) [inline plugins/API/API.php]: This long method needs a refactoring one day :) not related to your PR :)

**Comment by tsteur** [comment]
I don't understand everything but looks ok to me :) Are there any changes re performance or so? Of course using the CustomVariableName/Page segment will be slower and I think the changes in AutoSuggestAPI also only affect this segment right?

File: plugins/API/API.php (hunk 3 above)
Comment (diosmosis) [inline plugins/API/API.php]: I noticed many methods needing refactoring :) I even added a TODO for the first one. Then I saw the rest and lost hope :)

**Comment by diosmosis** [comment]
> It is still possible to apply a segment for a specific one right?

Yes, still possible. The other ones are added through an event.

> Can we maybe show a title in the UI that says that we're search in all custom variables for one segment and that results can be inaccurate if the search matches values from different custom variables? If I understand this feature right one has to be very careful using it, maybe even allowing only == comparison? 

This should only be a problem if doing a segment like `customVariableValue=...`. I don't think anyone will normally do this, favoring instead `customVariableName=...;customVariableValue=...`. Or just `customVariableName=...`.

> Are there any changes re performance or so?

There shouldn't be, I tried to make the old behavior the default in every change. I was most worried about accidental BC breaks.

> Of course using the CustomVariableName/Page segment will be slower and I think the changes in AutoSuggestAPI also only affect this segment right?

That's the idea. My changes should always use the fastest code path. The changes to AutoSuggestAPI will partially affect any segment that uses `suggestedValuesCallback`, but the extra work involved is just creating a `ReflectionMethod/ReflectionFunction` instance, which I think is trivial.

**Comment by diosmosis** [comment]
@mattab This is waiting for review and/or merge.

File: tests/PHPUnit/System/expected/test_apiGetReportMetadata__API.getSegmentsMetadata.xml [hunk 4]
```diff
@@ -279,6 +279,12 @@
 	<row>
 		<type>dimension</type>
 		<category>Custom Variables</category>
+		<name>Custom Variable name</name>
```
Comment (mattab) [inline tests/PHPUnit/System/expected/test_apiGetReportMetadata__API.getSegmentsMetadata.xml]: should be `Custom Variable name (scope visit)`

File: tests/PHPUnit/System/expected/test_apiGetReportMetadata__API.getSegmentsMetadata.xml [hunk 5]
```diff
@@ -339,6 +351,12 @@
 	<row>
 		<type>dimension</type>
 		<category>Custom Variables</category>
+		<name>Custom Variable value (scope page)</name>
+		<segment>customVariablePageValue</segment>
+	</row>
+	<row>
+		<type>dimension</type>
+		<category>Custom Variables</category>
```
Comment (mattab) [inline tests/PHPUnit/System/expected/test_apiGetReportMetadata__API.getSegmentsMetadata.xml]: should be `Custom Variable value (scope page)`

**Comment by mattab** [comment]
Good to merge after fixing the Segment names!

LGTM

File: tests/PHPUnit/System/expected/test_apiGetReportMetadata__API.getSegmentsMetadata.xml (hunk 5 above)
Comment (diosmosis) [inline tests/PHPUnit/System/expected/test_apiGetReportMetadata__API.getSegmentsMetadata.xml]: What should? The category?

Comment (mattab) [inline tests/PHPUnit/System/expected/test_apiGetReportMetadata__API.getSegmentsMetadata.xml]: My bad, this one is already good

**Pull request merged**